In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
M = pd.read_csv('movies.csv', index_col='Name')

In [4]:
M.head()

American Sniper  The Hunger Games: Mockingjay - Part 1  \
Name                                                                         
Aarti Jaiswal                     4                                    NaN   
Aditya Dharmasagar                5                                    NaN   
Adrian Cavallaris               NaN                                      4   
Ahmed Muheebuddin                 5                                    NaN   
Aishwarya reddy                 NaN                                      3   

                    Guardians of the Galaxy  The Lego Movie  The Hobbit  \
Name                                                                      
Aarti Jaiswal                           NaN             NaN           4   
Aditya Dharmasagar                      NaN               4           5   
Adrian Cavallaris                         3             NaN           4   
Ahmed Muheebuddin                         4             NaN           3   
Aishwarya reddy                           4               3           4   

                    Transformers  Malificent  Big Hero 6  Godzilla  \
Name                                                                 
Aarti Jaiswal                  3         NaN           5       NaN   
Aditya Dharmasagar             4           4         NaN         4   
Adrian Cavallaris              2         NaN         NaN       NaN   
Ahmed Muheebuddin            NaN         NaN         NaN       NaN   
Aishwarya reddy                4           3         NaN         4   

                    Interstellar  How to Train your Dragon 2  Gone Girl  \
Name                                                                      
Aarti Jaiswal                NaN                         NaN          4   
Aditya Dharmasagar             5                           5          5   
Adrian Cavallaris            NaN                         NaN        NaN   
Ahmed Muheebuddin            NaN                         NaN        NaN   
Aishwarya reddy                4                           3          5   

                    Divergent  The Fault in Our Stars  Unbroken  \
Name                                                              
Aarti Jaiswal             NaN                     NaN       NaN   
Aditya Dharmasagar          4                       4       NaN   
Adrian Cavallaris           4                     NaN       NaN   
Ahmed Muheebuddin         NaN                     NaN       NaN   
Aishwarya reddy           NaN                       4       NaN   

                    300: Rise of an Empire  
Name                                        
Aarti Jaiswal                          NaN  
Aditya Dharmasagar                       5  
Adrian Cavallaris                      NaN  
Ahmed Muheebuddin                      NaN  
Aishwarya reddy                          3

In [8]:
M.shape

(76, 16)

In [9]:
M

American Sniper  \
Name                                           
Aarti Jaiswal                       4.000000   
Aditya Dharmasagar                  5.000000   
Adrian Cavallaris                        NaN   
Ahmed Muheebuddin                   5.000000   
Aishwarya reddy                          NaN   
Akhilesh                                 NaN   
Andrew Webb                         4.000000   
Anirudh                             3.000000   
Anirudh Thota                            NaN   
anurag                              4.000000   
Anusha Putta                             NaN   
Arun Vedere                         4.000000   
Ashutosh Sureka                     5.000000   
Aswini kumar                        4.000000   
azam                                4.000000   
Benjamin Baysinger                  5.000000   
Bhuvana Posina                      5.000000   
Bradley Brubaker                         NaN   
Brant Mullinix                           NaN   
Charlie                                  NaN   
David Huang                         5.000000   
David Reyling                            NaN   
Evan Piercy                         4.000000   
Graham Bullard                      3.000000   
Greg Werner                         5.000000   
Harish Kallianpur                   5.000000   
Heather Sanders                          NaN   
HJ SONG                                  NaN   
J A Barnes                               NaN   
Jake Metzger                             NaN   
...                                      ...   
sai bhargav musuluri                3.000000   
Sai Sampath                              NaN   
Sai Yallapragada                    5.000000   
saikiran reddy                      3.000000   
sandeep katragadda                  5.000000   
Sara Collins                             NaN   
Shannon Morrison                         NaN   
Shiva rama raju Chekuri             4.000000   
siddanth sudarshan                  5.000000   
Sina Tuy                                 NaN   
solomon choppara                    5.000000   
Sundar Prasad                            NaN   
Sunil Nandikanti                         NaN   
surya chandra reddy kovvuri              NaN   
Susan Huang                              NaN   
Suyendra                                 NaN   
Syed Abbas Hussaini                 4.000000   
Syed Farzan                              NaN   
Thomas Podlipni                     5.000000   
Trudy Kline                         3.000000   
Tyler Vitanova                           NaN   
ugesh reddy challa                       NaN   
veerendra battula                   3.000000   
Venkatesh Komma                          NaN   
Violeta Lee Fay                          NaN   
Yinka Adeyemo                       4.000000   
Sai Sharan Vellala                  5.000000   
NaN                               159.000000   
NaN                                38.000000   
NaN                                 4.184211   

                             The Hunger Games: Mockingjay - Part 1  \
Name                                                                 
Aarti Jaiswal                                                  NaN   
Aditya Dharmasagar                                             NaN   
Adrian Cavallaris                                                4   
Ahmed Muheebuddin                                              NaN   
Aishwarya reddy                                                  3   
Akhilesh                                                       NaN   
Andrew Webb                                                      3   
Anirudh                                                          4   
Anirudh Thota                                                    3   
anurag                                                         NaN   
Anusha Putta                                                   NaN   
Arun Vedere                                                      3   
Ashutosh Sureka         

In [10]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

In [11]:
pearson(M['The Fault in Our Stars'],M['Malificent'] )

0.39873118100855826

In [12]:
pearson(M['The Fault in Our Stars'], M['American Sniper'])

-0.0027509195890370419

In [15]:
def get_recs(movie_name, M, num):

    import numpy as np
    
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

In [16]:
recs = get_recs('The Fault in Our Stars', M, 10)

In [17]:
recs[:10]

[('Malificent', 0.39873118100855826),
 ('Godzilla', 0.27910171761590286),
 ('How to Train your Dragon 2', 0.22514201090703723),
 ('Gone Girl', 0.14705909007675133),
 ('Interstellar', 0.12286970596222707),
 ('Unbroken', 0.11298392969275164),
 ('Guardians of the Galaxy', 0.090874959467711777),
 ('Divergent', 0.071129882767397401),
 ('The Hunger Games: Mockingjay - Part 1', 0.053533966431183633),
 ('The Lego Movie', 0.051121037070687873)]

In [18]:
anti_recs = get_recs('The Fault in Our Stars', M, 8551)

In [19]:
anti_recs[-10:]

[('Unbroken', 0.11298392969275164),
 ('Guardians of the Galaxy', 0.090874959467711777),
 ('Divergent', 0.071129882767397401),
 ('The Hunger Games: Mockingjay - Part 1', 0.053533966431183633),
 ('The Lego Movie', 0.051121037070687873),
 ('Transformers', 0.042958080479063222),
 ('The Hobbit', 0.038628454138809519),
 ('American Sniper', -0.0027509195890370419),
 ('300: Rise of an Empire', -0.053185970909404523),
 ('Big Hero 6', -0.18665349454812516)]

# 1Ans: Maleficient is the most similar movie to "The fault in our stars" as per the above outputs that we have received.

In [21]:
recs = get_recs('Malificent', M, 10)

In [22]:
recs[:1]

[('The Fault in Our Stars', 0.39873118100855826)]

In [25]:
ratings = M.ix['Sara Collins'].dropna()

In [29]:
ratings.sort(ascending=False)
fav_movie = ratings.index[0]

C:\Users\shru.shru-PC\Anaconda2\lib\site-packages\ipykernel\__main__.py:1: FutureWarning: sort is deprecated, use sort_values(inplace=True) for for INPLACE sorting
  if __name__ == '__main__':


In [26]:
fav_movie = ratings.index[0]

In [27]:
recs = get_recs('Gone Girl', M, 10)

In [28]:
recs[:10]

[('The Fault in Our Stars', 0.14705909007675133),
 ('The Lego Movie', 0.1233839307397927),
 ('How to Train your Dragon 2', 0.11123592989213005),
 ('Godzilla', 0.095217741548246887),
 ('300: Rise of an Empire', 0.087816187972314844),
 ('Unbroken', 0.070310485168203027),
 ('Interstellar', 0.054486961644112257),
 ('American Sniper', 0.021549485282159685),
 ('Divergent', 0.014708426410739026),
 ('Transformers', 0.0058415580617180805)]

# 2Ans: By using the get recs function and passing my favourite movie as "Gone Girl" and requesting for 10 recommendations as per the review given by my classmates. Given below are my 10 recommendations:
('The Fault in Our Stars', 0.14705909007675133),
 ('The Lego Movie', 0.1233839307397927),
 ('How to Train your Dragon 2', 0.11123592989213005),
 ('Godzilla', 0.095217741548246887),
 ('300: Rise of an Empire', 0.087816187972314844),
 ('Unbroken', 0.070310485168203027),
 ('Interstellar', 0.054486961644112257),
 ('American Sniper', 0.021549485282159685),
 ('Divergent', 0.014708426410739026),
 ('Transformers', 0.0058415580617180805)


# 3Ans: According to the recommendations that I've received from my classmates reviews, I would rate the movie(s) that I haven't seen as per the highest positive coefficient number between my favourite movie and the movie that I haven't watched.